### PR TITLE
New: Airstream, experimental state propagation library

### PR DESCRIPTION
**Pull request by raquo**
Airstream is a simple event stream and state propagation libraries with a concept of ownership and no FRP glitches. This PR defines the initial version of Airstream and integrates it into Laminar. This will be the next version of Laminar.

# Now with docs!

**See [Airstream README](https://github.com/raquo/Laminar/blob/airstream/src/main/scala/com/raquo/laminar/experimental/airstream/README.md
)**

---

#### Note: description below is really outdated. Airstream is evolving faster than I can document it right now...

Took a shot at the simplest, most limited state propagation library that I can imagine over the weekend.

Design goals:
- Represent UI state in a simple way
- Strong focus on absolute state consistency
- 100% Predictable execution (what, and in which order) without redundant calculations
- TODO: Implicit external contexts manage lifecycle of "subscriptions" (that term is not well defined yet).

Features:
- Var-s are Signal-s that can be updated externally
- ComputedSignal-s are Signal-s that depend on other Signals
- You create ComputedSignal-s using .map or Signal.combine
- You can add observers to Signal-s using Signal.foreach

Propagation of changes works approximately like this: depth-first propagation until you meet a Signal that depends on more than one Signal. At that point, mark this signal as pending, and don't try to evaluate it or any of its dependencies. When this is all done, see which of pending signals don't depend on other pending signals, and continue propagating that one. Rinse and repeat that last stage until there are no pending signals left. There are more details in the comments in Airstream.scala file.

Properties:
- Everything happens synchronously
- Definition: Propagation is a set of Signal updates that begins when you update one or more Var-s, and ends when all Signals that (directly or indirectly) depend on the updated Var-s were updated.
- Signals are recalculated EXACTLY once per propagation
- Signals are ONLY recalculated if at least one of their parents' values has updated AND changed
- Signals are recalculated AFTER all their dependencies that needed to be recalculated have finished recalculating
- Signals are recalculated regardless of any observers defined on any of them
- Observers are fired AFTER the propagation has completed, in the order in which the Signals that they are observing were updated during the propagation.
- Observers are fired once per their Signal update (i.e. once per propagation)
- Triggering a Var update inside an observer will start a new propagation that will complete before the rest of observations get to run. When this new propagation completes, any new observations that need to run will be added to the back of the list of pending observations, and executed accordingly.

Limitations
- State must be an **acyclic** graph. Any cycles must be created outside of the propagation (e.g. using observers or async streams) such that cyclical state updates are not triggered during a propagation
- Only Scala.js for now (target environment that also happens to be single-threaded)
- There are probably more that I forgot or didn't think of yet

Not implemented yet
- Error propagation
~~- Memory management (Context)~~
- Interface with streams
- Some performance improvements are possible but for a proof of concept I don't care yet

### CODE REVIEW

**Comment by raquo** [comment]
After today's experimentation with the concept of Owners, I think I'll actually see this through. I have a good plan on how to implement Streams and how to tie them together with Signals.

This is not going to be a general purpose library. It will be rather opinionated, optimized for a use case similar to Laminar: self-contained, hierarchical UI components, with each component owning signals and subscriptions defined "inside" of its body. So the end user will never need to kill any signals or subscriptions manually when writing UI code, and there will be a type safe distinction between signals (state) and event streams.

While the library should hopefully end up well suited for its task, it will also be quite limited in features and constraining in usage (on purpose, I have an ideology to push), and its size in terms of LoC will probably be an order of magnitude smaller than e.g. Monix.

Curious to see how it turns out, definitely one of the more interesting things I've tried.

**Comment by fdietze** [comment]
I'm happy you did this. I have a lot to read and understand now... 👍

**Comment by raquo** [comment]
Streams were... remarkably simple to implement, after a long search for the right structure. It's not perfect yet – there are many refactors ahead, especially to converge the functionality with Signals. And need to actually test the Streams to weed out stupid mistakes.

As I'm going through this, I'm getting a better idea of the conceptual differences and similarities between state, streams, and streams-with-initial-value. It's too much to write up, but in all likelihood big changes are coming to the original concept of Signals. Current implementation of Signals is very conservative in terms of strictly ensuring consistency. It is also likely to be somewhat costly in big applications (resolution of pending signals specifically). I have a plan to address that.

**Comment by doofin** [comment]
What is the disadvantage of xstream?

**Comment by raquo** [comment]
@doofin Xstream is a nice small streaming library, but:

- XStream suffers from FRP glitches, for example the diamond case glitch #2.
- XStream has (almost) no typeful distinction between Signals (time-varying values) and Event streams. You basically can't tell if a particular "stream" is guaranteed to have an initial value or not.
- Due to both of the above, XStream is quite unsuitable for modelling **state** (see [this comment](https://github.com/OutWatch/outwatch/issues/96#issuecomment-346654224) for extra details). Airstream solves those problems.

Also:
- XStream only has lazy observables – that is, they are only evaluated if someone subscribes to them. But component state should really be evaluated regardless of whether anyone is listening right now, otherwise it could be inconsistent. Airstream has State for this purpose.
- XStream has no concept of "components owning resources". For example, that means that if you manually create a subscription inside a component, and then the component gets unmounted, that subscription continues to live, leaking memory and wrecking havoc. Both outwatch and laminar have ways to create subscriptions bound to the component's lifetime, but neither safeguards the user against creating subscription manually using the streaming library directly. Airstream makes that impossible, as you **need** to specify an owner to every leaky resource you create (including subscriptions).

Almost all of the above problems are unsolved not only in XStream, but also for Rx, Monix, and other streaming libraries, which is why I'm making Airstream.

**Comment by raquo** [comment]
[airstream status update]

I'm currently reworking how synchronization (mechanism that prevents glitches) works. I'm going back to my original plan of having it always-on. It will be better to have safety on by default, with manual overrides possible for cases when performance is critical and the developer is willing to promise that they're not combining related events in a subset of the dependency  graph.

Additionally, the new setup will use transactions to fix unexpected "event-swallowing" behaviour in case of MergeEventStream in cyclical dependency graphs. And will also fix the order of events coming out of MergeEventStream in case of a glitch.

After this new system is done I will work on FlattenEventStream (`streamOfStreams.flatten()`) and then on integration between events, signals, and state. Everything's already planned out, just needs to be executed.

I'll document all this once it actually works. For now, just noting that I'm alive and making steady progress, I'm just busy with other stuff most days.

**Comment by doofin** [comment]
Great ! Actually you are making progress too fast that i can't keep up with haha.

**Comment by raquo** [comment]
I made a major change in how synchronization works. Now using topological order computed ahead-of-time at no cost to decide which pending CombineObservable-s to resolve first.

Previously Airstream needed to determine which of the pending observables were free of dependencies on other pending observables, and that had an unpleasant O-complexity. That old solution was an attempt to deal more gracefully with dependency cycles, but ultimately it wasn't worth it, especially because Airstream is designed for unidirectional dataflow apps. Cycles are still supported, but FRP glitches are possible if you were to do weird things with cycles (there really is no universal way around that).

The main benefit of this change is improved performance when propagating long & wide dependency graphs with a lot of CombineObservable-s. The synchronization code is also easier to understnand now – much more obvious when a new transaction will be triggered.

Speaking of transactions: Airstream guarantees absence of FRP glitches inside a transaction. When it's not possible or not desirable, a new transaction is created for subsequent propagation after the current transaction is done.

**Comment by raquo** [comment]
[Airstream status update]

I finally wired together EventStreams, Signals and State – there are now methods to create these from each other. Of course, added a bunch of tests for all this as well. I think I've validated the idea and the mechanics as much as possible in a vacuum, now I will start to integrate all this into Laminar. Yays!

Airstream code currently stands at only 1100 LoC (1400 with comments and empty lines), with about the same amount of LoC in tests. So, pretty good: 20% more than Scala.rx (but we do provide event streams, not just state), 3x smaller than XStream.js (but we're missing some operators, and those are JS lines), and 20x smaller than the monix-reactive package (which is a whole different beast, but just for reference).

Airstream will still grow a bit to include error handling and more operators but it's still respectably small, which I care about a lot for frontend code, and for simplicity in general.

**Comment by raquo** [comment]
[Airstream status update]

I have been really busy with non-Laminar work the last couple weeks. I am 95% done integrating Airstream into Laminar. It went very smooth, the only hiccups were around subscription lifecycle management (predictably) and mount event architecture (due to XStream's asynchronous unsubscribe).

There is still some work to do, including updating Laminar's documentation with Airstream, but I really hope to wrap this up over the weekend.

**Comment by raquo** [comment]
🎉 **[OMG Airstream Docs!](https://github.com/raquo/Laminar/blob/542aeeb1f337e8c89a640fdbddc19b8e1c6199ec/src/main/scala/com/raquo/laminar/experimental/airstream/README.md)** 🎉

🎉 **Airstream is now in Laminar!** 🎉

See commit message in https://github.com/raquo/Laminar/pull/8/commits/542aeeb1f337e8c89a640fdbddc19b8e1c6199ec for notable Laminar API changes. Apart from XStream and Airstream differences, there's almost nothing.

Note: This is very early, but with some changes this will eventually become the next version of Laminar. Laminar docs still need to be updated to reflect all the XStream -> Airstream changes.

File: build.sbt [hunk 1]
```diff
@@ -4,7 +4,6 @@ enablePlugins(ScalaJSBundlerPlugin)
 
 libraryDependencies ++= Seq(
   "org.scala-js" %%% "scalajs-dom" % "0.9.4",
-  "com.raquo.xstream" %%% "xstream" % "0.3.1",
```
Comment (raquo) [inline build.sbt]: 🎉

Comment (doofin) [reply build.sbt]: Wow!

**Comment by doofin** [comment]
I agree that Airstream focus on scala js only , for that there are already akka-stream (although I haven't tried it yet )

**Comment by raquo** [comment]
Well, the changes I've been doing recently are more of individual Airstream or Laminar improvements rather than integration code for those, so I guess now is as good a time as ever to wrap up this humongous PR.

Subsequent development towards v0.3 will take place either by me pushing directly to master (yolo) or in other individual PRs. And there's always #1 for general status updates.

**Pull request merged**